src/env.rs

```diff
@@ -0,0 +1,40 @@
+use std::net::SocketAddr;
+use std::sync::Arc;
+use std::time::Duration;
+
+use anyhow::Error;
+use slog::Logger;
+use tokio::sync::Mutex;
+
+use crate::client::RpcClient;
+use crate::fetch_blocks::{PeerHandle, PeerList};
+use crate::users::Users;
+
+#[derive(Debug)]
+pub struct TorEnv {
+    pub proxy: SocketAddr,
+    pub only: bool,
+}
+
+#[derive(Debug)]
+pub struct Env {
+    pub bind: SocketAddr,
+    pub rpc_client: RpcClient,
+    pub tor: Option<TorEnv>,
+    pub users: Users,
+    pub logger: Logger,
+    pub peer_timeout: Duration,
+    pub peers: Mutex<PeerList>,
+    pub max_peer_age: Duration,
+}
+impl Env {
+    pub fn leak(self) -> &'static Self {
+        Box::leak(Box::new(self))
+    }
+    pub fn arc(self) -> Arc<Self> {
+        Arc::new(self)
+    }
+    pub async fn get_peers(&self) -> Result<Vec<PeerHandle>, Error> {
+        self.peers.lock().await.get_peers(self).await
```

Comment:
This pattern is notoriously known for leading into various kinds of deadlocks. A better approach is:

```rust
struct Env {
// ...
peers: RwLock<Arc<PeerList>>,
}
```

then to get the peers:

```rust
let peers = self.peers.lock_read().clone();
```

When peers need to be updated, create a new collection from scratch and swap them:

```rust
std::mem::swap(&mut *self.peers.lock_write(), &mut new_peers);
```

In order to avoid multiple tasks from re-fetching the peer list, have a separate task that periodically calls into `bitcoind` and fills `peers`. This will also avoid clients from waiting on the operation to finish. The previous value of `peers` can be kept to attempt to reuse the allocation (`Arc::get_mut()`) in this case. To make sure some sane value exists in `peers` this operation can be forced at start with blocking.

Reply:
since I'm using an `mpsc` to cache a TCP connection on each peer, unfortunately I need a `&mut Peer` to access it. If I switched to a `mpmc`, it would allow this, but I'm not super aware of the tradeoffs.